utxoscanner.go

```diff
@@ -0,0 +1,410 @@
+package neutrino
+
+import (
+	"container/heap"
+	"sync"
+	"sync/atomic"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcwallet/waddrmgr"
+)
+
+// reportAndError is a simple pair type holding a spend report and error.
+type reportAndError struct {
+	report *SpendReport
+	err    error
+}
+
+// GetUtxoRequest is a request to scan for OutPoint from the height StartHeight.
+type GetUtxoRequest struct {
+	// OutPoint is the target outpoint to watch for spentness.
+	OutPoint *wire.OutPoint
+
+	// StartHeight is the height at which we expect to find the original
+	// unspent outpoint. This is also the height used when starting the
+	// search for spends.
+	StartHeight uint32
+
+	// resultErrChan either the spend report or error for this request.
+	resultErrChan chan reportAndError
+
+	// result caches the first spend report or error returned for this
+	// request.
+	result *reportAndError
+
+	// mu ensures the first response delivered via resultErrChan is in fact
+	// what gets cached in result.
+	mu sync.Mutex
+
+	quit chan struct{}
+}
+
+// deliver tries to deliver the report or error to any subscribers. If
+// resultErrChan cannot accept a new update, this method will not block.
+func (r *GetUtxoRequest) deliver(report *SpendReport, err error) {
+	select {
+	case r.resultErrChan <- reportAndError{report, err}:
```

Comment:
Yeah, this is definitely more defensive than is necessary if we are 100% certain there are no bugs :) With the previous logic, failing to deliver a ntfn would only block that GetUtxo call. With the batched version, the entire pipeline would grind to a halt if such a bug was introduced.

I've added a warning to the `default` case, so that we can be alerted to any occurrences in the wild.